### PR TITLE
Provide information on how S3MapDataset works

### PR DESCRIPTION
**Pull request by IsaevIlya**
## Description
Add information about how S3MapDataset is able to support random access pattern. Explain that S3MapDataset has to build the list of all objects to support random access. This process could be resource demanding, so it would start only when user try to access the actual data. 

--------
By submitting this pull request, I confirm that my contribution is made under the terms of BSD 3-Clause License and I agree to the terms of the [LICENSE](https://github.com/awslabs/s3-connector-for-pytorch/blob/main/LICENSE).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -52,17 +52,26 @@ End to end example of how to use `s3torchconnector` can be found under the
 
 The simplest way to use the S3 Connector for PyTorch is to construct a dataset, either a map-style or iterable-style 
 dataset, by specifying an S3 URI (a bucket and optional prefix) and the region the bucket is located in:
-```shell
+```py
```
Comment (dnanuti) [inline README.md:55]: The IDE suggests `python` when trying to update this section.

File: README.md [hunk 2]
```diff
@@ -52,17 +52,26 @@ End to end example of how to use `s3torchconnector` can be found under the
...
 iterable_dataset = S3IterableDataset.from_prefix(DATASET_URI, region=REGION)
 
+# Datasets are also iterators. 
+for object in iterable_dataset:
+  print(object.key)
+
+# S3MapDataset eagerly lists all the objects under the given prefix 
+# to provide support of random access.  
+# S3MapDataset build list of all objects at the first access to its elements or 
+# at the first call to get count of its elements, whichever happens first.
```
Comment (dnanuti) [inline README.md:71]: Should we mention directly `len` here?

File: README.md (hunk 1 above)
Comment (IsaevIlya) [reply README.md:55]: I have tried to turn on rich view, it works with just `py`

File: README.md (hunk 2 above)
Comment (IsaevIlya) [reply README.md:71]: I want to have first part and second part of sentence to be written in the same manner. To be both descriptive or both of them using code.

File: README.md [hunk 3]
```diff
@@ -52,17 +52,26 @@ End to end example of how to use `s3torchconnector` can be found under the
...
 # You need to update <BUCKET> and <PREFIX>
 DATASET_URI="s3://<BUCKET>/<PREFIX>"
 REGION = "us-east-1"
 
-map_dataset = S3MapDataset.from_prefix(DATASET_URI, region=REGION)
-
 iterable_dataset = S3IterableDataset.from_prefix(DATASET_URI, region=REGION)
 
+# Datasets are also iterators. 
+for object in iterable_dataset:
```
Comment (muddyfish) [inline README.md:65]: In this example, can we use `obj`, as `object` is a builtin and shadowing it isn't a good practice

**Review by muddyfish — APPROVED** [review]

**Review by dnanuti — APPROVED** [review]

**Pull request merged**